### PR TITLE
parser: reject closing delimiters that look like an operator

### PR DESCRIPTION
**Pull request by Kijewski**
Related issue: #128.

### CODE REVIEW

**Comment by Kijewski** [comment]
@ronnodas, do you think that the error message in this PR is good enough?

File: rinja_derive/src/config.rs [hunk 1]
```diff
@@ -360,6 +360,26 @@ impl<'a> RawSyntax<'a> {
             }
         }
 
+        for end in [syntax.block_start, syntax.expr_start, syntax.comment_start] {
```
Comment (GuillaumeGomez) [inline rinja_derive/src/config.rs:363]: Unless I'm missing something, you never check for closing delimiters here, right?

**Comment by ronnodas** [comment]
Yes, the error message looks good but it seems the code is comparing the starting delimiters even though it maybe should be comparing the ending ones.

File: rinja_derive/src/config.rs [hunk 2]
```diff
@@ -360,6 +360,26 @@ impl<'a> RawSyntax<'a> {
...
+            for prefix in ["<<", ">>", "&&", "..", "||"] {
+                if end.starts_with(prefix) {
+                    let msg = if end == prefix {
+                        format!("a closing delimiter must not start with an operator: {end:?}")
```
Comment (GuillaumeGomez) [inline rinja_derive/src/config.rs:367]: It'd be nice to make the distinction between closing and opening delimiter (in the current code, it's only opening delimiters).

File: rinja_derive/src/config.rs [hunk 3]
```diff
@@ -360,6 +360,26 @@ impl<'a> RawSyntax<'a> {
...
+                    } else {
+                        format!(
+                            "a closing delimiter must not start an with operator: \
+                             {end:?} vs {prefix:?}",
```
Comment (GuillaumeGomez) [inline rinja_derive/src/config.rs:371]: The "vs" part is a bit weird. I'd reword it like: `a closing/opening delimiter must not start with an operator: {end:?} starts with {prefix:?}`.

**Comment by GuillaumeGomez** [comment]
> Yes, the error message looks good but it seems the code is comparing the starting delimiters even though it maybe should be comparing the ending ones.

I mentioned the same thing 😄

**Comment by Kijewski** [comment]
Hehe, yeah, I should look what I copy and paste. Now the terminators are checked and the message is reworded.

File: rinja_derive/src/config.rs [hunk 4]
```diff
@@ -360,6 +360,26 @@ impl<'a> RawSyntax<'a> {
             }
         }
 
+        for end in [syntax.block_end, syntax.expr_end, syntax.comment_end] {
```
Comment (GuillaumeGomez) [inline rinja_derive/src/config.rs:363]: I think it should be checked for both starters and enders, not just enders (or starters in the previous version).

Comment (Kijewski) [reply rinja_derive/src/config.rs:363]: It's not needed, because we do not try to understand text blocks. The parsing will work with any starting delimiter. I would not like to introduce restrictions needlessly.

Comment (GuillaumeGomez) [reply rinja_derive/src/config.rs:363]: Fair enough. :+1:

File: rinja_derive/src/config.rs [hunk 5]
```diff
@@ -360,6 +360,26 @@ impl<'a> RawSyntax<'a> {
...
+            for prefix in ["<<", ">>", "&&", "..", "||"] {
+                if end.starts_with(prefix) {
+                    let msg = if end == prefix {
+                        format!("a closing delimiter must not start with an operator: {end:?}")
```
Comment (GuillaumeGomez) [inline rinja_derive/src/config.rs:367]: nitpicking: it would be nice to know which block is impacted by this. Ie "expr_end" would be "expression block". So in this case the message would look like:

```
a closing delimiter must not start with an operator. Expression block closing delimiter is {end:?} which is also an operator
```

File: rinja_derive/src/config.rs [hunk 6]
```diff
@@ -360,6 +360,26 @@ impl<'a> RawSyntax<'a> {
...
+                    } else {
+                        format!(
+                            "a closing delimiter must not start an with operator: \
+                             {end:?} starts with {prefix:?}",
```
Comment (GuillaumeGomez) [inline rinja_derive/src/config.rs:371]: Same here with a little tweak:

```
a closing delimiter must not start with an operator. Expression block closing delimiter ({end:?}) starts with {prefix:?}
```

Comment (Kijewski) [reply rinja_derive/src/config.rs]: Makes sense! I changed the text in the other delimiter tests, too.

File: rinja_derive/src/config.rs [hunk 7]
```diff
@@ -321,45 +321,92 @@ impl<'a> RawSyntax<'a> {
...
 
-        for s in [
-            syntax.block_start,
-            syntax.block_end,
-            syntax.expr_start,
-            syntax.expr_end,
-            syntax.comment_start,
-            syntax.comment_end,
+        for (s, k) in [
+            (syntax.block_start, "opening block"),
```
Comment (GuillaumeGomez) [inline rinja_derive/src/config.rs:325]: Good idea implementing it here too!

File: rinja_derive/src/config.rs [hunk 8]
```diff
@@ -321,45 +321,92 @@ impl<'a> RawSyntax<'a> {
...
+                if end.starts_with(prefix) {
+                    let msg = if end == prefix {
+                        format!(
+                            "a closing delimiter must not start with an operator. \
+                             The {kind} delimiter ({end:?}) is also an operator",
+                        )
+                    } else {
+                        format!(
+                            "a closing delimiter must not start an with operator. \
+                             The {kind} delimiter ({end:?}) starts with {prefix:?}",
```
Comment (GuillaumeGomez) [inline rinja_derive/src/config.rs:398]:
```suggestion
                             The {kind} delimiter ({end:?}) starts with the {prefix:?} operator",
```

File: rinja_derive/src/config.rs [hunk 9]
```diff
@@ -321,45 +321,92 @@ impl<'a> RawSyntax<'a> {
...
+            (syntax.expr_start, "opening expression"),
+            (syntax.expr_end, "closing expression"),
+            (syntax.comment_start, "opening comment"),
+            (syntax.comment_end, "closing comment"),
         ] {
             if s.len() < 2 {
                 return Err(CompileError::new_with_span(
-                    format!("delimiters must be at least two characters long: {s:?}"),
+                    format!(
+                        "delimiters must be at least two characters long. \
```
Comment (GuillaumeGomez) [inline rinja_derive/src/config.rs:335]: I realize we don't seem to have ui tests for this. I'll send a PR that.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks!

**Comment by Kijewski** [comment]
Thank you for the reviews and the good idea to call out the offending delimiter.

@ronnodas, thank you for reporting the bug!

File: rinja_derive/src/config.rs (hunk 9 above)
Comment (GuillaumeGomez) [reply rinja_derive/src/config.rs:335]: Nevermind, tests are at the bottom.
